### DISCUSSION TITLE
Ghost Spectre - How to create social network and graph ?

### DISCUSSION THREAD

**Discussion by vramULB** [cmu-sei/GHOSTS#385] [Q&A]
I have created some NPCs and clients thanks to the new version of the project, but I'm unsure how to use the Spectre functionalities.

I noticed that the documentation refers to configurations based on the old version of the application, and the current Spectre functionalities were integrated into the Ghosts API.

I found the new configuration in the appsettings.json of the Ghosts API Docker container, which indicates that the Spectre functionality should work.

![image](https://github.com/user-attachments/assets/b0c5bc0c-d736-4ae8-aad7-071ab6c056d1)

**Comment by sei-dupdyke** [comment]
These are animator functions. You can find some documentation on these [here](https://cmu-sei.github.io/GHOSTS/animator/) and setting up the API generally [here](https://cmu-sei.github.io/GHOSTS/core/api/).

Basically, once you have the API up and running, you should be able to go to the Animations tab and start these jobs. SocialGraph is the simplest and should just run, so please try that one first. This will build relationships and knowledge share between agents randomly.

![image](https://github.com/user-attachments/assets/3a3468e5-055b-4dbc-b47f-eb2d4fed715d)

You can pull and look at this information from the API with 

GET /api/npcs/{id} Get NPC by specific Id

There is still lots to do in order to make this information more directly integrated with say, influencing new activities — like you and I spoke at the coffee counter about X, so then I went back to my desk and searched for X on the internet, etc. But it is there for scripting at least for now.

HTH

**Reply by vramULB** [reply]
Thank you for the explanation.
I tried to run the social graph on Animation tab, but it seems not to work. Maybe I'm doing something wrong.

 -   I generated 5 Linux VMs using the same username as the NPCs below. I synchronized to verify that the usernames correspond to the machines: `/api/npcsgenerate/syncwithmachineusernames`. Is this the correct way to link machines with NPCs?

![image](https://github.com/user-attachments/assets/417aa93d-81b4-4714-93a7-339fa9ebe491)


- I ran basics timelines as BrowserFirefox and Commands which work correctly
- Then, when I tried to run the animation job Social Graph, it create an instance but it seems to run in the void.
![image](https://github.com/user-attachments/assets/039d34e7-95b1-4e73-8e55-b795b4a9054a)
- I checked the Ghosts API logs. They show me that the social graph is being processed, but the ID seems weird.
![Screenshot from 2024-08-05 15-20-44](https://github.com/user-attachments/assets/23f0d1c8-9e26-4b05-8b13-de351b734817)
-  I also checked the NPC information which contain now a SocialGraph but it has empty values.
```
...
  "npcSocialGraph": {
    "id": "00000000-0000-0000-0000-000000000000",
    "name": null,
    "connections": [],
    "knowledge": [],
    "beliefs": null,
    "currentStep": 0
  }
```

**Reply by sei-dupdyke** [reply]
Animator should probably disclaim _Here be dragons_ at the top ;)

So, on the page that lists all your users, that should mean you have NPCs built, available in the `npcs` table in the database, and those should return in the api /api/npcs endpoint. 

Can you confirm that none of their IDs are "0000-..."?

**Reply by devdevg00se** [reply]
I'm experiencing the same issue. I have generated 10 NPCs, run the Social Graph job via the web app and all the SocialGraphJob logs are like 

`2024/08/08 13:19:54.607|ghosts.api.Infrastructure.Animations.AnimationDefinitions.SocialGraphJob.Step|7: 00000000-0000-0000-0000-000000000000 is interacting...
`

The Social Graph tab in the web app still produces an error (I guess the graph is not generated). 

Hitting the api/npcs/list endpoint returns the following data:

```
[
 {
    "id": "01c0e41b-0b93-421e-815e-2f2f1020277d",
    "name": "Beale Lariviere"
  },
  {
    "id": "108a05ae-fab1-4aa8-9af1-dd56affddc1f",
    "name": "Thom Waytowich"
  },
  {
    "id": "176bf208-6e21-4ceb-bcc3-a922aabaaa5e",
    "name": "Donnie McGruder"
  },
  {
    "id": "59f86f48-b4bf-4d60-b4be-c21787904ea3",
    "name": "Rubina Kidd"
  },
  {
    "id": "5f6c7fa8-4902-4433-b50d-af9ccb4ecc66",
    "name": "Chantalle Yum"
  },
  {
    "id": "7735ad8f-5925-4715-a5f6-2b30b8c3452e",
    "name": "Alwin De Boer"
  },
  {
    "id": "a2e909d3-c2ed-43ba-b2c9-41d2754db2eb",
    "name": "Hadrian Whitty"
  },
  {
    "id": "a679ddfd-ce41-4b77-bb2b-761e90083ee0",
    "name": "Lawerence Chatel"
  },
  {
    "id": "a84064b4-d8f4-483b-84b8-f1fb964f2032",
    "name": "Gabriellia Dingman"
  },
  {
    "id": "ceca0b3a-e80f-4f0b-9a84-3a469cead754",
    "name": "Pail Burchat"
  }
]
```

**Comment by sei-dupdyke** [comment]
Definitely a problem, give me a bit to troubleshoot here locally. Sorry for the inconvenience.

**Comment by sei-dupdyke** [comment]
I posted up a temporary fix #391 for this in the latest docker image (8.1.4):

`docker pull dustinupdyke/ghosts:8.1.4`
